### PR TITLE
OpenPGP 3.4+: Enable ed25519/curve25519 support

### PR DESCRIPTION
**Pull request by ya-isakov**
This MR add support of 25519-series of curves to OpenSC. It extends support added in https://github.com/OpenSC/OpenSC/pull/1960.

Tested with Yubikey 5 NFC, Firmware 5.2.4.

It includes fix for Yubikey, which returns extra 0 at the end of OID in Algorithm Attribute DO. I've changed sc_compare_oid to return in it found -1 in any OID. This way, all trailing bytes will be ignored for comparison, including Import-Format.

p11test works, the only issue is with test_usages, which complains that derive should not be set on public key. I think, this should be an issue for all ECDH supporting tokens, as it's set in https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15-openpgp.c#L480 and https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15-openpgp.c#L484.

##### Checklist

- [x] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

P.S. Right now I'm just checking if version of card is more or equal to 3.4, but it seems, that there is a better way to detect list of supported algorithms, via Algorithm Information DO added in spec version 3.4. I've addressed that in https://github.com/OpenSC/OpenSC/pull/2287

### CODE REVIEW

**Comment by dengert** [comment]
In regards to: "derive should not be set on public key." I disagree. Other drivers in OpenSC set
the usage for ECDH pub key to derive. (It is not used directly, but the peer uses it as input for a derive operation. 

I can not find directly in PKCS11 if it is appropriate or not. But an EC private key can be used for Sign ECDSA or derive ECDH. X509 cert also have usage and PKCS15 also has usage.
If used for Sign the public key has a usage of verify. 
My argument is if a certificate or the private key has a usage of derive, this fact should also be expressed in the usage of the public key.  

Policies may require some EC keys to be used only for signatures and others may only be used for derivation. If you don't allow for a usage of derive on public key,  you can not tell from the public key if the private key is allowed for derive.

We need a PKCS11 or other authoritative reference.

**Comment by mouse07410** [comment]
Of course, @dengert is right. What idiot wrote a test that declares invalid "derive" for ECDH keys?! What Key Usage would _you_ assign to such a key?

File: src/libopensc/card-openpgp.c
Comment (Jakuje) [inline src/libopensc/card-openpgp.c:101]: copy&paste comment which no longer reflects the reality.

File: src/libopensc/card-openpgp.c
Comment (Jakuje) [inline src/libopensc/card-openpgp.c:113]: This is the same as the original `ec_curves_openpgp` aside of the curve25519 and ed25519 addition so I am wondering if we could simplify this somehow to avoid list duplication. Either by prepending these new to the original list and using `ec_curves_openpgp` as a pointer to the third element or some other creative (but readable) way?

File: src/libopensc/card-openpgp.c
Comment (ya-isakov) [reply src/libopensc/card-openpgp.c:101]: Thank you! And what is reality now? I thought that these RFCs are still valid

File: src/libopensc/card-openpgp.c
Comment (ya-isakov) [reply src/libopensc/card-openpgp.c:113]: Thank you! I'm changing it to be similar to a pgp_objects

**Review by Jakuje — APPROVED** [review]

**Comment by ya-isakov** [comment]
Appveyor problems are not related to my changes, it seems, but I don't know how to retrigger it...

**Comment by Jakuje** [comment]
> Of course, @dengert is right. What idiot wrote a test that declares invalid "derive" for ECDH keys?! What Key Usage would _you_ assign to such a key?

Contributions and reviews are always welcomed.

**Comment by ya-isakov** [comment]
I'm reworking OID comparison, as I don't like the idea of changing the function, which worked before. While researching on it, I've found a bug in parsing of last byte in Algorithm Attribute - if it's 00, which is a mark that no pubkey import is supported, it;s not ignored, and parsed as a part of oid. While fixing that, I realized that parse of Algorithm Attributes in DO 6E should be done, even if DO FA was parsed, as in Yubikey, DO FA entries has no last byte set to FF or 00 at all. So, to be on the safe side, I'm adding the logic that will parse DO 6E, and if algo was added previously, it will update it's keyformat.

Not a fan of this solution, but it's better than to make some wrong assumptions.

**Comment by ya-isakov** [comment]
So, I reverted change of OID comparison logic, and instead, strip trailing zero byte, in case if it was read from DOs. Same logic as is used by gnupg (https://github.com/gpg/gnupg/blob/1c16878efd0bcf036f56abef93d64ac41ce9e95b/scd/app-openpgp.c#L5969-L5971 and https://github.com/gpg/gnupg/blob/1c16878efd0bcf036f56abef93d64ac41ce9e95b/scd/app-openpgp.c#L5969-L5971)
This logic was in place, but it ignored last byte only when it was 0xFF, and this was checked only in reading Algorithm Attributes while parsing list of supported curves, but same DO is read while parsing Public/Private keys, so I've added it there.

**Comment by ya-isakov** [comment]
@Jakuje I'm happy with algorithm now, so, should I wait for someone with merge rights to merge it?

**Pull request merged**
